Refactor with torchmetrics

KnathanM opened the pull request:
(Note: this PR is based on #1020 and should be merged after it. #1024 is based on this PR.)

This PR seeks to reduce developer burden by switching our loss functions and metrics to `torchmetrics`. This has the following benefits:

* `torchmetrics` is developed by the `lightning` team so they are made to work together, e.g. automatic synchronization of metrics during distributed training. 
* `torchmetrics` provides metrics we don't have. It is straightforward to adapt these to work with our `MPNN` which means we can point people to `torchmetrics` when they want more loss functions/metrics.
* Validation and Test metrics are no longer batch averaged.

This has the following challenges:
* A predictor's `criterion` is saved as a whole object in a model file. This makes it difficult to unpickle if we rename and move class definitions. See below for my work around.
* Classification metrics need to save all predictions and targets in order to calculate metrics like AUROC. This takes more memory than batch averaging the metrics. I don't know if the extra memory used is a problem.

Adapting our loss functions/metrics to `torchmetrics.Metric`s was pretty direct. The `forward` method of `LossFunction` got split into separate `update` and `compute` methods. `Metric` has three key methods. `__init__` sets up internal states that are used to calculate the metric. In MSE for example, these are the squared error (`self.total_loss`) and total number of samples (`self.num_samples`). `update` takes the preds and targets and updates the internal states. `compute` uses the internal states to calculate the final metric. The `forward` method of `torchmetrics.Metric`s is a bit more complicated. It will 
1. save the state of the metric instance
2. reset the metric
3. call `update` and `compute` using the batch to give the loss value for that batch
4. reload the original state of the instance and incorporate the state of the batch's `update` call. 

We use `forward` during training so we can get gradients. During validation we don't need that so we can just call `update`. Then `self.log` will call `compute` to get the metric value to report. 

### Custom unpickle
* The loss functions got moved to `nn.metrics` and some of them got renamed. The unpickler doesn't know where to find the class definition because the location it saved in the model file is no longer there. I found I could intercept what gets passed to the unpickler's `find_class` method by making a custom unpickle class in a custom pickle module. Not sure how stable or advisable that is, but it works. 
* Even after the unpickler manages to unpickle the criterion saved in old checkpoint files, those criterion are not initialized properly as `torchmetrics` which give an error about the criterion not have the attribute `_defaults`. I get around this by reinitializing the criterion if needed.
* I only noted the name changes for loss functions because the metrics never gave me trouble. I think this is because the metrics weren't saved as children previously. But now that I think of it, probably if a metric was passed into `MPNN`, it would be saved in the hyperparameters and need to be unpickled. I guess all of our checkpoint files in the repo just used `metrics=None`. I'll let this PR go through a round of reviews before I address this. 

### MCC 
This metric is included in `torchmetrics` but that version doesn't allow for task or data weights, so I didn't replace ours. I did have to modify it a bit though to accommodate the ability to weight tasks differently. `torchmetrics` initializes the metric's states to some default, usually a tensor, in which case, the size of the tensor needs to be known when initialized. We don't require users to say how many tasks they have however. So instead I set the default to an empty list which can hold any size tensor. This is less memory efficient, but means less code changes. 

### Multiclass metrics
* I assume that `CrossEntropyMetric` was the default metric for `MulticlassClassficationFFN` because we didn't really have other multiclass metrics available. [Online](https://stats.stackexchange.com/questions/370861/why-is-cross-entropy-not-a-common-evaluation-metric-for-model-performance) is seems that people don't use CE as a metric. I switched this to use `MulticlassMCCMetric`.
* To make `MulticlassMCCMetric` more accurate, I added another `1 - MCCLoss` so that a value of 1 does indicate a perfect model. 
* Multiclass metrics are difficult to implement in the CLI because in `torchmetrics` the number of classes is required when creating the metric instance, but we dynamically get the number of classes from the shape of `preds`. Either more multiclass metrics should to be manually added to chemprop using our style, or we should to adapt the CLI to pass the number of classes to the metric, or we can require people to use a jupyter notebook if they want a different metric for multiclass. I think generally if users want metrics that aren't in Chemprop, including other multiclass metrics, it is reasonable to ask them to use a jupyter notebook. In the docs notebooks for loss functions and metrics, I added an example of adapting multiclass metrics from `torchmetrics` to work with chemprop, to be a resource we can point users to.

### Train loss on epoch
I added `on_epoch=True` to calculate the training loss over the whole epoch because otherwise the train loss reported is only for a single batch of data and it may be much higher or lower than average, which may be confusing for users. Calculating the loss at the end of the epoch requires accumulating the preds/targets for all training batches though, which may be a problem for large training sets... I haven't tested large datasets. But also, I'm not sure if setting `on_epoch=False` in `self.log` in `training_step` stops the accumulation anyways. 


### Tests
We are using `torchmetrics` correctly if changing the batch size and shuffling the data doesn't change the metric calculation for a dataset. I test this in `test_metrics.py`. This file has example targets and predictions for every predictor for both `train_step` and `forward`. It then fake trains a model and captures the output of the progress bar to check that "train_loss_epoch" equals "val_loss" and that the validation metric equals the test metric. 
* This also ensures that all of the loss functions/metrics don't have run time errors. 
* It does not test that the metrics are computed correctly. 

`test_loss_functions.py` tests if some metrics are computed correctly.
| Metric | Tested? |
| --- | --- |
| MSE | Implicitly by testing BoundedMSE | 
| MAE | Not tested, seems trivial? |
| RMSE | Not tested but just MSE with a `sqrt()` |
| BoundedMSE | Tested |
| BoundedMAE | The bounded part is tested by BoundedMSE | 
| BoundedRMSE | " " |
| R2Score | Inherited from `torchmetrics` which tests it |
| MVELoss | Tested |
| Evidential | Tested |
| BCELoss | Tested |
| CrossEntropyLoss | Tested |
| BinaryMCCLoss | Tested | 
| BinaryMCCMetric | Implicitly by BinaryMCCLoss |
| MulticlassMCCLoss | Tested | 
| MulticlassMCCMetric | Implicitly by MulticlassMCCLoss |
| BinaryAUROC | Inherited from `torchmetrics` which tests it |
| BinaryAUPRC | Not tested, I have manually checked it before... |
| BinaryAccuracy | Inherited from `torchmetrics` which tests it |
| BinaryF1Score | Inherited from `torchmetrics` which tests it |
| BinaryDirichletLoss | Tested | 
| MulticlassDirichletLoss | Tested |
| SID | Tested |
| Wasserstein | Tested |


### Reviewing
I recommend looking at this PR commit by commit as I spent a while trying to logically group changes together. 

Also note that I only updated `validate_loss_function` enough to get the code to run. I plan to remove that function in #1023 so I didn't both making it actually good.

JacksonBurns commented on chemprop/cli/train.py:
```diff
@@ -469,11 +478,16 @@ def process_train_args(args: Namespace) -> Namespace:
+    if args.tracking_metric not in valid_tracking_metrics:
+        raise ArgumentError(
+            argument=None,
+            message=f"Tracking metric must be either 'val_loss' or one of the metrics specified via `--metrics`. Got {args.tracking_metric}",
```
Perhaps instead:
```python
message=f"Tracking metric must be one of {','.join(valid_tracking_metrics)}. Got {args.tracking_metric}"
```
to just show them what their options are.

JacksonBurns commented on chemprop/cli/utils/utils.py:
```diff
@@ -34,11 +34,11 @@ def _pop_attr_d(o: object, attr: str, default: Any | None = None) -> Any | None:
     match predictor_ffn:
         case predictors.RegressionFFN:
-            if criterion not in (loss.MSELoss, loss.BoundedMSELoss):
+            if criterion not in (loss.MSE):
```
If you still want this to be a tuple, it needs to be `(loss.MSE, )`

KnathanM replied:
This is removed after merging v2-1/dev into this branch.

JacksonBurns commented on chemprop/models/model.py:
```diff
@@ -91,9 +90,9 @@ def __init__(
 
         self.metrics = (
-            nn.ModuleList([*metrics, self.criterion])
+            nn.ModuleList([*metrics, self.criterion.clone()])
```
Why clone here?

KnathanM replied:
From the [docs](https://lightning.ai/docs/torchmetrics/latest/pages/overview.html):
"Metrics contain internal states that keep track of the data seen so far. Do not mix metric states across training, validation and testing. It is highly recommended to re-initialize the metric per mode as shown in the examples above."
I use `.clone()` to get a new instance of the metric used for the training loss.

JacksonBurns commented on chemprop/models/model.py:
```diff
@@ -167,17 +166,11 @@ def validation_step(self, batch: TrainingBatch, batch_idx: int = 0):
-            prog_bar=True,
-            sync_dist=distributed.is_initialized(),
-        )
+        self.metrics[-1](preds, targets, mask, weights, lt_mask, gt_mask)
```
This line does nothing (?)

JacksonBurns replied:
I don't think this affects readability enough to be worth changing

KnathanM replied:
Yes it does, which is why torchmetrics recommends doing it this way.

JacksonBurns commented on chemprop/nn/loss.py:
```diff
@@ -2,32 +2,33 @@
-class LossFunction(nn.Module):
-    minimize: bool = True
+class ChempropMetric(torchmetrics.Metric):
+    is_differentiable = True
```
If we are using this attribute just to decide if we can use a given `ChempropMetric` for training or not, I would rather this be called `supports_training` or something like that.

JacksonBurns replied:
I'm aware this is a significant change - please feel free to push back if this is just semantics

KnathanM replied:
`is_differentiable` is an [attribute](https://lightning.ai/docs/torchmetrics/stable/pages/implement.html#metric-attributes) of `torchmetrics.Metric`, and I am not sure where in their code it is used.

JacksonBurns replied:
Interesting... ¯\_(ツ)_/¯

Nevermind on the rename, of course

JacksonBurns commented on .flake8:
```diff
@@ -6,3 +6,4 @@ per-file-ignores =
     __init__.py: F401
     chemprop/nn/predictors.py: F405
     chemprop/nn/metrics.py: F405
+    tests/unit/nn/test_metrics.py: E121, E122, E131, E241, W291
```
How bad is the reformat if we remove this? Like, how many lines?

JacksonBurns reviewed (changes requested):
Thanks for taking on this monster PR - much appreciated.

Overall in great shape. Some small notes throughout.

> Classification metrics need to save all predictions and targets in order to calculate metrics like AUROC. This takes more memory than batch averaging the metrics. I don't know if the extra memory used is a problem.

I don't think we need to worry about this, but good to keep in mind.

RE: pickle problems

Let's say we didn't implement this Pickle fix. If my understanding is correct, models trained in 2.0 would not be loadable in 2.1 (if they used one of the renamed metrics)?

If that's the case, I am thinking maybe we move the pickle fix into a small 2.0to2.1 script, similar to the v1tov2 script. It would do what you have here, but only require users to run it once, rather than every time during inference (and require us to maintain this code in the future).

KnathanM commented:
I like the pickle fix script idea. Originally I felt it was more slick to handle it without the user needing to do anything, but now I worry having the pickle conversion in the main part of the code will complicate it for future readers. I'll see if I can get it in a separate script. Some of the checkpoint files in `tests/data` will need to be regenerated.

KnathanM commented:
I have now moved the pickle stuff to its own file. I also updated the checkpoint files using
`for i in *pt; do python ../../chemprop/utils/v2_0_to_v2_1.py $i $i; done` while in `tests/data`.

I found that the reaction and spectra checkpoint files were already out of date (they aren't included in `tests/regenerate_models.sh`), so I deleted them.

JacksonBurns reviewed (approved):
Looks sound - please merge at your leisure.

shihchengli commented on chemprop/nn/metrics.py:
```diff
@@ -63,160 +77,473 @@ def forward(
         lt_mask: Tensor | None = None,
         gt_mask: Tensor | None = None,
     ):
+        """Calculate the mean loss function value given predicted and target values
+
+        Parameters
+        ----------
+        preds : Tensor
+            a tensor of shape `b x t x u` (regression), `b x t` (binary classification), or
+            `b x t x c` (multiclass classification) containing the predictions, where `b` is the
+            batch size, `t` is the number of tasks to predict, `u` is the number of
+            targets to predict for each task, and `c` is the number of classes.
+        targets : Tensor
+            a float tensor of shape `b x t` containing the target values
+        mask : Tensor
+            a boolean tensor of shape `b x t` indicating whether the given prediction should be
+            included in the loss calculation
+        weights : Tensor
+            a tensor of shape `b` or `b x 1` containing the per-sample weight
+        lt_mask: Tensor
+        gt_mask: Tensor
+
+        Returns
+        -------
+        Tensor
+            a scalar containing the fully reduced loss
+        """
```
1. add  `-> None`
2. remove the docstring related to return
```suggestion
    ) -> None:
        """Calculate the mean loss function value given predicted and target values

        Parameters
        ----------
        preds : Tensor
            a tensor of shape `b x t x u` (regression), `b x t` (binary classification), or
            `b x t x c` (multiclass classification) containing the predictions, where `b` is the
            batch size, `t` is the number of tasks to predict, `u` is the number of
            targets to predict for each task, and `c` is the number of classes.
        targets : Tensor
            a float tensor of shape `b x t` containing the target values
        mask : Tensor
            a boolean tensor of shape `b x t` indicating whether the given prediction should be
            included in the loss calculation
        weights : Tensor
            a tensor of shape `b` or `b x 1` containing the per-sample weight
        lt_mask: Tensor
        gt_mask: Tensor
        """
```

shihchengli commented on chemprop/nn/metrics.py:
```diff
@@ -63,160 +77,473 @@ def forward(
+        task_weights :  ArrayLike, default=1.0
+            the per-task weights of shape `t` or `1 x t`. Defaults to all tasks having a weight of 1.
+        """
+        super().__init__()
+        task_weights = torch.as_tensor(task_weights, dtype=torch.float).view(1, -1)
+        self.register_buffer("task_weights", task_weights)
```
```suggestion
        super().__init__(task_weights)
```
If that makes sense, please change the rest of them

shihchengli commented on chemprop/nn/metrics.py:
```diff
@@ -63,160 +77,473 @@ def forward(
+        return 1 - super().compute()
+
+
+class ChempropClassificationMixin:
```
```suggestion
class ClassificationMixin:
```

shihchengli replied:
Also, all the `Mixin` functions are missed in `__init__.py`

KnathanM replied:
I took inspiration from classes like `_MolGraphDatasetMixin`. They are no supposed to be imported by users so they are not included in the init.py file, but can still be imported directly when needed. (`from chemprop.data.datasets import _MolGraphDatasetMixin`). I will add a `_` to the name of the mixin to suggest that.

shihchengli commented on chemprop/utils/v2_0_to_v2_1.py:
```diff
@@ -0,0 +1,40 @@
+
+
+if __name__ == "__main__":
+    import v2_0_to_v2_1
+
+    model = torch.load(sys.argv[1], map_location="cpu", pickle_module=v2_0_to_v2_1)
```
```suggestion
    import sys

    model = torch.load(sys.argv[1], map_location="cpu", pickle_module=sys.modules[__name__])
```

shihchengli commented on docs/source/tutorial/python/metrics.ipynb:
```diff
@@ -290,7 +312,297 @@
+   "source": [
+    "Chemprop metrics are instances of `chemprop.nn.metrics.ChempropMetric`, which inherits from `torchmetrics.Metric`. Custom loss functions need to follow the interface of both `ChempropMetric` and `Metric`. Start with a `Metric` either by importing an existing one from `torchmetrics` or by creating your own by following the instructions on the `torchmetrics` website. Then make the following ch ...
+    "\n",
+    "\"1. Allow for task weights to be passed to the `__init__` method.\n",
+    "\"2. Allow for the `update` method to be given `preds, targets, mask, weights, lt_mask, gt_mask` in that order.\n",
+    "\"3. Provide an alias property, which is used to identify the metric value in the logs.\n",
```
```suggestion
    "1. Allow for task weights to be passed to the `__init__` method.\n",
    "2. Allow for the `update` method to be given `preds, targets, mask, weights, lt_mask, gt_mask` in that order.\n",
    "3. Provide an alias property, which is used to identify the metric value in the logs.\n",
```

KnathanM commented on chemprop/nn/metrics.py:
```diff
@@ -1,60 +1,74 @@
-            Ignored. Maintained for compatibility with :class:`~chemprop.nn.loss.LossFunction`
-    """
+class ChempropMetric(torchmetrics.Metric):
+    is_differentiable = True
+    higher_is_better = False
+    full_state_update = False
```
torchmetrics says these are optional but recommended.

KnathanM commented on chemprop/cli/train.py:
```diff
@@ -535,7 +535,9 @@ def validate_train_args(args):
-            message=f"Tracking metric must be either 'val_loss' or one of the metrics specified via `--metrics`. Got {args.tracking_metric}",
+            message=f"Tracking metric must be one of {','.join(valid_tracking_metrics)}. "
+            f"Got {args.tracking_metric}. Additional tracking metric options can be specified with "
+            "the --metrics flag.",
```
This suggestion is to make the message all on one line, is that right? I prefer to split the string over multiple lines so that each line isn't too long. But happy to change that to one line if you'd like.

shihchengli replied:
I am find to keep multiple lines here. The reason I made this suggestion is because the rest of the help messages in this module are written in one line.

KnathanM commented:
@shihchengli The main thing I had to fix during merging v2-1/dev in was adding the Quantile loss. Can you check that again to make sure it is how you expect? I added it to metrics.py and the `__init__.py` file. I also added a note that we should add a test for it in `test_loss_functions.py` at some point in the future.

shihchengli reviewed (approved):
The change looks good to me
